cpp/src/parquet/column_writer.cc

```diff
@@ -2305,6 +2307,74 @@ struct SerializeFunctor<
   int64_t* scratch;
 };
 
+// ----------------------------------------------------------------------
+// Write Arrow to Float16
+
+// Requires a custom serializer because Float16s in Parquet are stored as a 2-byte
+// (little-endian) FLBA, whereas in Arrow they're a native `uint16_t`. Also, a temporary
+// buffer is needed if there's an endian mismatch.
+template <>
+struct SerializeFunctor<::parquet::FLBAType, ::arrow::HalfFloatType> {
+  Status Serialize(const ::arrow::HalfFloatArray& array, ArrowWriteContext* ctx,
+                   FLBA* out) {
+#if ARROW_LITTLE_ENDIAN
+    return SerializeInPlace(array, ctx, out);
+#else
+    return SerializeWithScratch(array, ctx, out);
+#endif
+  }
+
+  Status SerializeInPlace(const ::arrow::HalfFloatArray& array, ArrowWriteContext*,
+                          FLBA* out) {
+    const uint16_t* values = array.raw_values();
+    if (array.null_count() == 0) {
+      for (int64_t i = 0; i < array.length(); ++i) {
+        out[i] = ToFLBA(&values[i]);
+      }
+    } else {
+      for (int64_t i = 0; i < array.length(); ++i) {
+        out[i] = array.IsValid(i) ? ToFLBA(&values[i]) : FLBA{};
+      }
+    }
+    return Status::OK();
+  }
+
+  Status SerializeWithScratch(const ::arrow::HalfFloatArray& array,
+                              ArrowWriteContext* ctx, FLBA* out) {
+    AllocateScratch(array, ctx);
+    if (array.null_count() == 0) {
+      for (int64_t i = 0; i < array.length(); ++i) {
+        out[i] = ToFLBA(array.Value(i));
+      }
+    } else {
+      for (int64_t i = 0; i < array.length(); ++i) {
+        out[i] = array.IsValid(i) ? ToFLBA(array.Value(i)) : FLBA{};
+      }
+    }
+    return Status::OK();
+  }
+
+ private:
+  FLBA ToFLBA(const uint16_t* value_ptr) const {
+    return FLBA{reinterpret_cast<const uint8_t*>(value_ptr)};
+  }
+  FLBA ToFLBA(uint16_t value) {
```

Comment:
The code look good to me, but the two mode here is a bit confusing, would different name for these `ToFLBA` better? (avoid mis-use the `ToFLBA` and touch `scratch_` )